### DISCUSSION TITLE
TCP Control Flow implemented?

### DISCUSSION THREAD

**Discussion by pmarin98** [omnetpp/omnetpp#1002] [Q&A] [answered]
Hi,
I am trying to enhance an existing TCP application for my Master Thesis and this is using the TCP protocol from the INET framework.
However, I observed that the flow control between the application and the TCP is not correctly implemented, since all the packets are from the receiver buffer are directly sent to the application. Therefore, the value of the advertised window from the receiver remains always the same.

This can be seen at the file inet/src/inet/transportlayer/tcp/TcpConnectionRcvSegment.cc, in lines 545-555, where there are actually two FIXME commented.

However, I observed from the release notes of INET-1.99.0 (March 3, 2011) that some functionality was added, at least experimentally.
In addition, I found a paper which tackles partially this problem (find attached) and there it says, the changes proposed by the authors were added.

Was actually the control flow already added? If not, do you have any suggestions how to implement it?
[OMNeTWorkshop2010-TCP.pdf](https://github.com/omnetpp/omnetpp/files/11905896/OMNeTWorkshop2010-TCP.pdf)

**Comment by avarga** [comment, marked as answer]
Indeed, flow control between the app and TCP is still not implemented. TCP immediately sends up all data it receives, so there's no way for the receiver app to slow down or block the sender side. Beyond the FIXME you mentioned in TcpConnectionRcvSegment.cc, I found some bits & pieces which would belong to the implementation: a TCP_C_READ command code, TCP_E_READ fsm event code, and the process_READ_REQUEST() method in TcpConnectionEventProc.cc. Unfortunately process_READ_REQUEST() doesn't even accept a "number of bytes" parameter. 

We aren't currently actively working on TCP, but if you are willing to take a look at it, we can assist you in that.

**Reply by pmarin98** [reply]
Okay, thank you very much for the reply. Yes, actually I would be willing to try to improve the functionality. I'll contact you.
